Relax type restrictions on null coalescing and null coalescing assign…

333fred opened the pull request:
…ment. @cston @dotnet/roslyn-compiler for review.

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -3361,15 +3362,7 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
                     Conversion.NoConversion, CreateErrorType(), hasErrors: true);
             }
 
-            // SPEC ERROR: The specification states:
```
The ECMA Spec for C#5 permits the `null` literal on the left-hand side of a null coalescing operator, so this comment is now unnecessary.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -3381,25 +3374,23 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
-            // SPEC ERROR: The spec does not call out that the left-hand side may not be a method group
-            // SPEC ERROR: or anonymous function. We should add a line to the spec to say so.
-
+            // SPEC: The left hand side must be either the null literal, or not a non-nullable value type. Method groups
```
`..., or a nullable value type.`

jaredpar replied:
What about type parameters?

---
In reply to: [213723498](https://github.com/dotnet/roslyn/pull/29577#discussion_r213723498) [](ancestors = 213723498)

cston replied:
Then perhaps `"..., or a nullable value type, or an unconstrained type parameter."`

---
In reply to: [213730337](https://github.com/dotnet/roslyn/pull/29577#discussion_r213730337) [](ancestors = 213730337,213723498)

333fred replied:
Both of these things are `not a non-nullable value type`

---
In reply to: [213757765](https://github.com/dotnet/roslyn/pull/29577#discussion_r213757765) [](ancestors = 213757765,213730337,213723498)

333fred replied:
@cston @jaredpar if you think the original wording is hard to read with the double negative, let me know. Otherwise, the existing wording covers these cases.

---
In reply to: [213770268](https://github.com/dotnet/roslyn/pull/29577#discussion_r213770268) [](ancestors = 213770268,213757765,213730337,213723498)

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:
```diff
@@ -47,14 +47,17 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
             // first we can make a small optimization:
             // If left is a constant then we already know whether it is null or not. If it is null then we 
```
>If left is a constant [](start = 15, length = 21)

Should this comment be moved to line 60?

333fred replied:
I moved it to above 54, as that's where the optimization starts.

---
In reply to: [213727642](https://github.com/dotnet/roslyn/pull/29577#discussion_r213727642) [](ancestors = 213727642)

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:
```diff
@@ -47,14 +47,17 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
             }
 
-            if (rewrittenLeft.ConstantValue != null)
+            if (rewrittenLeft.ConstantValue != null && !isUnconstrainedTypeParameter)
```
>isUnconstrainedTypeParameter [](start = 56, length = 28)

Is `isUnconstrainedTypeParameter` ever true when `ConstantValue != null`?

333fred replied:
Yes, you can have `default(T) ?? t1`.

---
In reply to: [213727960](https://github.com/dotnet/roslyn/pull/29577#discussion_r213727960) [](ancestors = 213727960)

jaredpar replied:
Won't `null` show up as a constant value?

333fred replied:
The constant value of the null literal is `ConstantValue.Null`, which is not actually `null`.

---
In reply to: [213847956](https://github.com/dotnet/roslyn/pull/29577#discussion_r213847956) [](ancestors = 213847956)

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:
```diff
@@ -47,14 +47,17 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
+            // it can be either left or right depending on the runtime type of T
 
-            if (rewrittenLeft.IsDefaultValue())
+            var isUnconstrainedTypeParameter = rewrittenLeft.Type?.IsReferenceType == false && rewrittenLeft.Type?.IsValueType == false;
```
Can `rewrittenLeft.Type` be `null` here? If not, consider replacing `?.` with `.`. If the type can be `null`, we only need to check once. (It may be clearer to check `rewrittenLeft.Type != null && …`.)

333fred replied:
Yeah, for the null literal. Will do.

---
In reply to: [213787027](https://github.com/dotnet/roslyn/pull/29577#discussion_r213787027) [](ancestors = 213787027)

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs:
```diff
@@ -1418,33 +1418,26 @@ public static void Main()
-                //         t ??= default;
-                Diagnostic(ErrorCode.ERR_BadBinaryOps, "t ??= default").WithArguments("??=", "T", "default").WithLocation(6, 9)
-            );
-
```
Consider leaving this test but with updated diagnostics.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs:
```diff
@@ -1418,33 +1418,26 @@ public static void Main()
+Assignment Evaluated
+0
+0
+");
```
Consider testing `where T : class` in addition to unconstrained.

333fred replied:
Also verified the behavior of `where T : struct`

---
In reply to: [213789722](https://github.com/dotnet/roslyn/pull/29577#discussion_r213789722) [](ancestors = 213789722)

AlekseyTs commented:
Should similar relaxation be done for VB?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs:
```diff
@@ -1418,33 +1418,26 @@ public static void Main()
     static void Main()
     {
         Verify<object>(null, ""Assignment Evaluated"");
+        Verify<int>(default, 10);
```
Consider adding a test with a nullable value type, with null and not null LHS.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:
```diff
@@ -1363,27 +1363,97 @@ .maxstack  2
+        Goo(default, 10);
+        Goo(1, 10);
+        Goo(default, ""String parameter 1"");
+        Goo(""String parameter 2"", ""Should not print"");
```
>Goo(""String parameter 2"", ""Should not print""); [](start = 8, length = 50)

Consider adding a test for a nullable value type

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:
```diff
@@ -1363,27 +1363,97 @@ .maxstack  2
+    static void Main()
+    {
+        Goo(10);
+        Goo(""String parameter"");
```
>Goo(""String parameter""); [](start = 8, length = 26)

Here as well.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs:
```diff
@@ -1418,33 +1418,26 @@ public static void Main()
-{
-    void M<T>(T t)
-    {
-        t ??= default;
```
>t ??= default [](start = 8, length = 13)

Is this a success case now? Do we have a test that verifies that the behavior (and perhaps IL) is what we expect?

AlekseyTs commented:
Done with review pass (iteration 2)

333fred commented:
VB LDM has been emailed, corresponding vblang issue: [https://github.com/dotnet/vblang/issues/339](https://github.com/dotnet/vblang/issues/339)

---
In reply to: [417062412](https://github.com/dotnet/roslyn/pull/29577#issuecomment-417062412) [](ancestors = 417062412)

333fred commented:
@AlekseyTs @cston addressed feedback.

---
In reply to: [417099139](https://github.com/dotnet/roslyn/pull/29577#issuecomment-417099139) [](ancestors = 417099139)

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:
```diff
@@ -44,17 +44,19 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
-            if (rewrittenLeft.IsDefaultValue())
+            // MakeConversion(left). This does not hold when the left is an unconstrained type parameter: at runtime,
+            // it can be either left or right depending on the runtime type of T
+            if (rewrittenLeft.IsDefaultValue() && !isUnconstrainedTypeParameter)
```
>!isUnconstrainedTypeParameter [](start = 50, length = 29)

It feels like it will be cheaper to check ```!isUnconstrainedTypeParameter``` first, and perhaps only once for the purpose of this ```if``` and the next one. Something like:
```
if (!isUnconstrainedTypeParameter)
{
            if (rewrittenLeft.IsDefaultValue())
            {
                return rewrittenRight;
            }

            if (rewrittenLeft.ConstantValue != null)
            {
                Debug.Assert(!rewrittenLeft.ConstantValue.IsNull);

                return GetConvertedLeftForNullCoalescingOperator(rewrittenLeft, leftConversion, rewrittenResultType);
            }
}
```

jcouv commented:
@333fred This language change should also be recorded somewhere, although it is done as part of `??=`.
We'll need to remember to document it (in VS release notes, language spec and MSDN docs).
